File: go/internal/sol/p0057/p0057.go
package p0057

/*
use recurrence relation:
  sqrt(2) = 1 + sqrt(2) - 1
          = 1 + 1 / ( 1 / (sqrt(2) - 1) )
          = 1 + 1 / ( (sqrt(2) + 1) / (2 - 1) )
          = 1 + 1 / (1 + sqrt(2))
  -->
  a{1} = 1 + 1/2
  a{n} = 1 + 1/(1 + a{n-1})    [n>1]

assume that b{n}/c{n} = a{n}
  b{1}/c{1} = 1 + 1/2 = 3/2
  b{n}/c{n} = 1 + 1/(1 + b{n-1}/c{n-1})
            = 1 + 1/((c{n-1) + b{n-1})/c{n-1})
            = 1 + c{n-1}/(c{n-1) + b{n-1})
            = (c{n-1) + b{n-1} + c{n-1))/(c{n-1) + b{n-1})
            = (2 * c{n-1} + b{n-1}) / (c{n-1) + b{n-1}
*/

import (
	"math/big"
	"strconv"
)

func compute(limit int) string {
	var result int
	b := big.NewInt(1)
	c := big.NewInt(1)

	for range limit {
		b.Add(c, b)
		c.Add(c, b)
		b, c = c, b
		if len(b.String()) > len(c.String()) {
			result++
		}
	}

	return strconv.FormatInt(int64(result), 10)
}

func Solve() string {
	return compute(1_000)
}

File: rust/src/bin/p0057.rs
// Project Euler: Problem 57

/*
  use recurrence relation:
    sqrt(2) = 1 + sqrt(2) - 1
            = 1 + 1 / ( 1 / (sqrt(2) - 1) )
            = 1 + 1 / ( (sqrt(2) + 1) / (2 - 1) )
            = 1 + 1 / (1 + sqrt(2))
    -->
    a{1} = 1 + 1/2
    a{n} = 1 + 1/(1 + a{n-1})    [n>1]

  assume that b{n}/c{n} = a{n}
    b{1}/c{1} = 1 + 1/2 = 3/2
    b{n}/c{n} = 1 + 1/(1 + b{n-1}/c{n-1})
              = 1 + 1/((c{n-1) + b{n-1})/c{n-1})
              = 1 + c{n-1}/(c{n-1) + b{n-1})
              = (c{n-1) + b{n-1} + c{n-1))/(c{n-1) + b{n-1})
              = (2 * c{n-1} + b{n-1}) / (c{n-1) + b{n-1})
*/

euler::run_solver!(57);

fn solve() -> String {
    compute(1_000).to_string()
}

fn compute(limit: i64) -> i64 {
    use num_bigint::BigUint;

    debug_assert!(limit > 0);

    let mut ans: i64 = 0;
    let (mut b, mut c) = (BigUint::from(1_u32), BigUint::from(1_u32));

    for _ in 0..limit {
        (b, c) = (&c + &c + &b, &c + &b);
        if BigUint::to_str_radix(&b, 10).len() > BigUint::to_str_radix(&c, 10).len() {
            ans += 1;
        }
    }

    ans
}

#[cfg(test)]
mod tests {
    use super::compute;

    #[test]
    fn p0057_8() {
        assert_eq!(compute(8), 1);
    }

    #[test]
    fn p0057_1000() {
        assert_eq!(compute(1_000), 153);
    }
}
